feat(linter): add linter service framework (#907)
canonical/craft-application: docs/how-to-guides/add-a-linter.rst

.. meta::
   :description: How to implement and integrate a linter in a craft app.

.. _how-to-add-a-new-linter:

Create a linter
===============

This guide shows how to add a linter to a craft app using the shared linter
service.

Implement the linter
--------------------

Create a new class that extends the ``AbstractLinter`` class and yields issues
in its ``run()`` method.

Use ``Stage.PRE`` when checks depend on source files and parsed project data
(``ctx.project``). Use ``Stage.POST`` when checks depend on packed output;
``ctx.artifact_dirs`` contains the extracted artifact directories to inspect.

Your ``run()`` method is your custom processing hook. Iterate the relevant
context data and yield ``LinterIssue`` entries when you find problems.

When linting artifacts in ``Stage.POST``:

* Iterate every directory in ``ctx.artifact_dirs`` instead of assuming a single artifact.
* Validate packaged content (for example, required files, layout, or metadata).
* Keep checks read-only and deterministic (avoid mutating files or relying on network access).

> 

Reviewer: ```suggestion

Here's an example linter that checks project and compiled part files:
```

Add a new linter
================

This guide shows how to add a linter to a *craft application* using the
shared linter service.

1) Implement a linter
---------------------

> Create a class that extends ``AbstractLinter`` and yields issues from ``run``.

Reviewer: ```suggestion
Create a new class that extends the ``AbstractLinter`` class and yields issues in its ``run()`` method.
```

Use ``Stage.PRE`` to lint the source tree (and access the parsed project via
``ctx.project``) and ``Stage.POST`` to lint packed artifacts via
``ctx.artifact_dirs``.

> .. code-block:: python

Reviewer: Please provide an example filename with:

```
    :caption: <path/to/file>.py
```

       name = "example.pre"
       stage = Stage.PRE

       def run(self, ctx: LintContext):
           # Inspect ctx.project_dir and/or ctx.artifact_dirs
           yield LinterIssue(
               id="E001",
               message="Example warning",
               severity=Severity.WARNING,
               filename=str(ctx.project_dir / "README.md"),
           )

2) Register the linter
----------------------

Register your linter class at import time:

.. code-block:: python

   from craft_application.services.linter import LinterService
   LinterService.register(MyPreLinter)

Ensure that the module containing your linter is imported when your
application starts (for example, from an app plugin or a module import).

3) Run the service
------------------

Use the service to stream issues for the desired stage:

.. code-block:: python

   from pathlib import Path
   from craft_application.lint import LintContext, Stage
   from craft_application.services.linter import LinterService

   svc = LinterService(app, services)
   svc.load_ignore_config(project_dir=Path.cwd())
   for issue in svc.run(Stage.PRE, LintContext(Path.cwd(), [])):
       print(issue)

>    print(int(svc.summary()))  # 0 OK, 1 WARN, 2 ERROR

Reviewer 2: I think copilot is confused.

4) Ignore configuration

> -----------------------

Reviewer: ```suggestion
Linter ignores by the user
--------------------------
```
